fix: enforce subtyping depth limit in WebAssembly module validation

deveshidwivedi opened the pull request:
This PR enforces the subtyping depth limit of 64 (as per the WebAssembly specification) during the validation of WebAssembly modules. Previously, WasmEdge did not enforce this limit, allowing modules with excessive subtyping depth to pass validation.
Changes:
- Added a maximum subtyping depth constant (`MAX_SUBTYPE_DEPTH` = 63).
- Implemented recursive subtype depth calculation with cycle detection.
- Added a validation check in the type validator to enforce depth constraints and prevent invalid subtyping structures.
Fixes: [#3949](https://github.com/WasmEdge/WasmEdge/issues/3949)

juntao commented:
Hello, I am a [code review agent](https://github.com/flows-network/github-pr-review/) on [flows.network](https://flows.network/).

It could take a few minutes for me to analyze this PR. Relax, grab a cup of coffee and check back later. Thanks!

hydai commented:
Hi @deveshidwivedi,  
Please also add a test case for this bug fix to ensure that it resolves the mentioned issue.

core-burner commented:
I had done this on my branch. Adding error type instead of using InvalidSubType would be more coherent for the user. It's just 2 lines 0x228.

dannypsnl commented on lib/validator/validator.cpp:
```diff
@@ -850,5 +861,27 @@ Expect<void> Validator::validateConstExpr(AST::InstrView Instrs,
+  const auto &Type = *TypeVec[TypeIdx];
+
+  for (auto SuperIdx : Type.getSuperTypeIndices()) {
+    if (SuperIdx < TypeVec.size()) {
```
If `SuperIdx >= TypeVec.size()`, does any error report needs to be made?

deveshidwivedi replied:
Yes, it fails validation with `ErrCode::Value::InvalidSubType`

deveshidwivedi commented on lib/validator/validator.cpp:
```diff
@@ -850,5 +861,27 @@ Expect<void> Validator::validateConstExpr(AST::InstrView Instrs,
   return Checker.validate(Instrs, Returns);
 }
 
+uint32_t Validator::calculateSubtypeDepth(uint32_t TypeIdx, std::set<uint32_t>& Visited) const {
```
updated

dannypsnl commented on include/validator/validator.h:
```diff
@@ -75,9 +76,13 @@ class Validator {
+
+  static constexpr uint32_t MAX_SUBTYPE_DEPTH = 63;
+
+  uint32_t calculateSubtypeDepth(uint32_t TypeIdx, std::set<uint32_t>& Visited) const;
```
The signature is mismatched, and `Depth` can be put to the final and with default value `0`, so caller has no need to know it's there.

deveshidwivedi replied:
Sorry for missing it earlier, I’ve updated the function signature.

dannypsnl commented on include/validator/validator.h:
```diff
@@ -75,9 +76,13 @@ class Validator {
   static inline const uint32_t LIMIT_MEMORYTYPE = 1U << 16;
   /// Proposal configure
-  const Configure Conf;
+  const Configure &Conf;
```
Does this change is required? This will turn a member to a reference, two semantics are very different.

deveshidwivedi replied:
I initially changed to `const Configure &Conf;` thinking it would be a performance optimization to avoid copying the `Configure` object.  However, I realize that this changes the semantics, as `Conf` is now a reference rather than a copy.

hydai commented on include/validator/formchecker.h:
```diff
@@ -54,7 +54,7 @@ class FormChecker {
 
   std::vector<VType> result() { return ValStack; }
-  auto &getTypes() { return Types; }
+  auto &getTypes() const { return Types; }
```
What's the reason for changing this function signature?

deveshidwivedi replied:
This is because, the `calculateSubtypeDepth()` method is a `const` method and it calls `Checker.getTypes()`, the `getTypes()` method must also be const to be callable from within a const context.

`auto &getTypes() { return Types; }` couldn't be called from the `calculateSubtypeDepth()` method.

dannypsnl replied:
I think the point is others place that invokes `getTypes()`, we would like to know their behaviour didn't get changed implicitly.

hydai commented on include/validator/validator.h:
```diff
@@ -78,6 +78,10 @@ class Validator {
+
+  static constexpr uint32_t MAX_SUBTYPE_DEPTH = 63;
+
+  Expect<void> calculateSubtypeDepth(uint32_t TypeIdx, std::set<uint32_t>& Visited, uint32_t Depth = 0) const;
```
Do we need to have this private function inside the class?
What if define it inside the details namespace of the implementation. We only need to pass the reference of TypeIdx and Checker into this function.

deveshidwivedi replied:
I understand, defined it inside the details namespace instead.

hydai commented on lib/validator/validator.cpp:
```diff
@@ -165,32 +165,38 @@ Expect<void> Validator::validate(const AST::SubType &Type) {
-  for (auto Index : Type.getSuperTypeIndices()) {
-    if (Index >= TypeVec.size()) {
+// Validate subtype depth when GC proposal is enabled
+if (Conf.hasProposal(Proposal::GC) && !Type.getSuperTypeIndices().empty()) {
```
You change the behavior when the GC is disabled. What's the expected behavior if we don't enable GC but the indices' size > 1

deveshidwivedi replied:
Updated to log an error, missed it in the previous implementation

deveshidwivedi commented:
Changes made:
- Moved `calculateSubtypeDepth` function to `details` namespace  
- Added error logging to handle cases where subtypes with multiple supertypes are invalid when GC is disabled
- Removed `const` from `FormChecker` to ensure expected behavior

Please let me know if the structure of code or any other aspect needs adjustment.

dannypsnl commented on include/validator/validator.h:
```diff
@@ -80,5 +80,9 @@ class Validator {
+
 } // namespace Validator
-} // namespace WasmEdge
+} // namespace WasmEdge
```
EOF

dannypsnl commented on include/validator/validator.h:
```diff
@@ -80,5 +80,9 @@ class Validator {
   FormChecker Checker;
 };
 
+namespace details {
+Expect<void> calculateSubtypeDepth(uint32_t TypeIdx, FormChecker &Checker);
+} // namespace details
```
This namespace no need to show in header actually.

deveshidwivedi replied:
When I ran the tests it said:
```code
/home/deveshi1/wasmedge/WasmEdge/lib/validator/validator.cpp:8:1: note: ‘std::set’ is defined in header ‘<set>’; did you forget to ‘#include <set>’?
````

deveshidwivedi replied:
Got it :)

dannypsnl commented on include/validator/validator.h:
```diff
@@ -81,4 +81,4 @@ class Validator {
 
 } // namespace Validator
-} // namespace WasmEdge
+} // namespace WasmEdge
```
EOF still there, maybe you forget to push the commit?

deveshidwivedi replied:
Apologies, updated now

dannypsnl reviewed (dismissed):
The big picture LGTM.

deveshidwivedi commented:
Hi @hydai @dannypsnl, I'm having trouble finding the right file to add tests for `validator.cpp`. Could you point me in the right direction?

hydai commented:
> Hi @hydai @dannypsnl, I'm having trouble finding the right file to add tests for `validator.cpp`. Could you point me in the right direction?

Hi @deveshidwivedi 
Due to the life time of the configuration and ASTModule, it will be easier to have this unit test inside the test/executor/executor.cpp.

dannypsnl commented:
@deveshidwivedi Had you run the tests locally? It seems not worked.

deveshidwivedi commented:
> @deveshidwivedi Had you run the tests locally? It seems not worked.

Yes I did run the tests locally, I'm trying to fix these currently.

hydai commented:
Hi @deveshidwivedi 
Are you going to work on this PR? If not, I will close it, thanks.

q82419 commented:
> Hi @deveshidwivedi Are you going to work on this PR? If not, I will close it, thanks.

@hydai As you mentioned the [spec](https://github.com/WebAssembly/gc/pull/360/files) in issue #3949 ,
This spec describes the limitation values of instance entries or sizes in V8.
The core spec allows runtimes to define their limitations, so this may not be considered as a bug.

hydai commented:
> @hydai As you mentioned the [spec](https://github.com/WebAssembly/gc/pull/360/files) in issue #3949 , This spec describes the limitation values of instance entries or sizes in V8. The core spec allows runtimes to define their limitations, so this may not be considered as a bug.

V8 and WebKit implement the same limitation. I think we should have either this limitation or a warning so that people can recognize when the program is stuck due to the subtyping depth issue.
